### PR TITLE
Merge #72, #59, #45, Refactoring and Theme System

### LINKED ISSUE #85: Some features from the original are not in this version
Can I change the color of wires or has this feature not been added? it is in the original.

### LINKED ISSUE #93: Is #79 ever gonna get approved
This is a cool merge request but it hasnt been touched in 6 months. please do something about #79

### PR DESCRIPTION
**Pull request by VitoBarra**
Merged https://github.com/DigitalLogicSimCommunity/Digital-Logic-Sim-CE/pull/45. Made some alterations to to fix bugs, including making it work with zoom, making it work with small .bin files, and changing the IO to busses for ease of use.
Merged https://github.com/DigitalLogicSimCommunity/Digital-Logic-Sim-CE/pull/59.
Merged https://github.com/DigitalLogicSimCommunity/Digital-Logic-Sim-CE/pull/72. Fixed 2 minor bugs.

In addition, I have done 

- Major refactoring of the project. (The important part of this PR :P)
- Added a Theme system for customizable wire and signal colors.
  - Change and Save\load Selected theme Color (Theme System completion)
  - Ability to Save\load Group size 
  - Ability to Save\load Signal WireType
- Enhanced the signal property menu to allow for group size and bus value changes by adding functionality and new keybindings: 
  - Q and E: to cycle between default themes.
  - R and F: Change the group size.
  - Save changes made in the GroupProperty menu using Enter and Space.
  - Discard changes made in the GroupProperty menu using Exit.
  - Use up arrow (+1) and down arrow (-1) to adjust integer input fields in the GroupProperty menu.

QOL of the main menu:
now in the load project menu you can 
- Delete a project
- Renominate a project
- Duplicate a project

Fix:
- issue #47 
- issue #48 and related #49 

### notes
To implement these features, the format of save files needed to be reworked, for now the format is not very efficient, more data than needed are stored for each subchip, this maybe will be fixed in the next release 

Credits haven't been updated

### CODE REVIEW

**Comment by crispeeweevile** [comment]
When will this PR be accepted? it's got a lot of cool features

**Comment by 901238746** [comment]
This is so cool, why hasn't it been approved yet?

**Comment by 901238746** [comment]
Man someone should approve this. Idk how to code but I dont think any of it is going to alt F4 the game

**Review by 901238746 — COMMENTED** [review]
Some of this looks fine but im a begginer. what happened to Palette.asset?

**Comment by 901238746** [comment]
I'd love this pull request because Im making a 11x16(11-bit 16 addresses) For my ALU and later I might need to make it 16 by 16 so that it would work with a better version of my ALUs LU(Logic Unit) so it can do more operations and comparisions. I just need more space. I made this in the original and I would hate it if I had to make it again.

**Review by 901238746 — APPROVED** [review]
This is fine

File: Assets/Modules/Save System/SaveSystem.cs
Comment (AOx0) [inline Assets/Modules/Save System/SaveSystem.cs:181]: I know you are actively working on the PR on your dev branch, just wanted to make you know about this one.

```cs
if (!Directory.Exists(oldAppDataPath)) return;
```

I believe the check does not actually make the cut; with a fresh build on a machine with a partial old path where the "Seb Lague" folder exists but not the "SaveData" (idk why) the code triggers an exception with the error:
```
DirectoryNotFoundException: Could not find a part of the path '/home/ae/.config/unity3d/Sebastian Lague/Digital Logic Sim/SaveData'.
```

However I think we can delay the check to validate `oldSaveDataPath` instead of `oldAppDataPath`. That way we verify the resource we are about to operate with actually exists and do not make assumptions based on the solely existence of "Seb Lague". 

This does work without triggering exceptions:

```diff
        string oldAppDataPath = Path.Combine(new string[]
        {
            Directory.GetParent(Application.persistentDataPath).Parent.FullName, "Sebastian Lague", "Digital Logic Sim"
        });
--      if (!Directory.Exists(oldAppDataPath)) return;

        string oldSaveDataPath = Path.Combine(oldAppDataPath, "SaveData");
++      if (!Directory.Exists(oldSaveDataPath)) return;
```

**Review by AOx0 — CHANGES_REQUESTED** [review]

File: Assets/Modules/Save System/SaveSystem.cs
Comment (VitoBarra) [reply Assets/Modules/Save System/SaveSystem.cs:181]: Thank you for letting me know, I’ll take a look!

**Comment by VitoBarra** [comment]
This should be good to go, some parts of the main menu are not very nice but they work

File: Assets/Scripts/UI/LoadProjectMenu.cs
Comment (AOx0) [inline Assets/Scripts/UI/LoadProjectMenu.cs]: `UnityEditor.EditorUtility` is not available on release builds. I don't think we have any solution for this in the project yet, only the `StandaloneFileBrowser` plugin to open/save dirs and files.

We could hide the button and leave it for future PRs, though.

**Review by AOx0 — CHANGES_REQUESTED** [review]

**Comment by AOx0** [comment]
Hey, I've been trying the PR, seems good in the Unity Editor, so far.
However, when producing a release build, the input system just stops working with pins, so wires don't work nor the activation/deactivation of the pins signal.

I suspect it's an issue with the new Unity Input system. I can reproduce the issue in the Unity Editor when changing the Event System components to the new suggested `Input System UI Input Module`.

My next thing on the list is to build the main branch's version and see if the issue is present there too, so I can discard the source code’s fault. Otherwise, if the main branch does work correctly, maybe it would be necessary to git bisect and revert whatever the change caused the button's fault. 

I don’t know if you are noticing the same on your end? It may be an issue with Unity in Linux.

File: Assets/Scripts/UI/LoadProjectMenu.cs
Comment (VitoBarra) [reply Assets/Scripts/UI/LoadProjectMenu.cs]: For now, I followed your advice and I have hidden the button

**Comment by VitoBarra** [comment]
It happens on my machine (Windows) as well.
I will work on it, but after a first investigation, I found out that issues arise when the SebInput Module is used. This module depends on the new input system, so probably needs to be updated or not used, as in the case of the signal handler that works fine.
I added a package (in the 3rdparty folder) that should be useful for debugging this problem.

**Comment by AOx0** [comment]
Thank you as always Vito! 

Just one little nitpick, I don't know if this is intentional, but the changing of pin colors does not display the correct color to be placed. i.e. It displays red but when placed the pin is green.

I'm willing to merge as-is and fix later tho, so I'll be just waiting for your confirmation whether to commit the merge or not.

**Comment by VitoBarra** [comment]
I'm not sure that I understand the concern, but it is intentional that selecting the color for signals (input/output chip pins) and cables requires separate actions.
With this said the PR should be ready c:

**Review by AOx0 — APPROVED** [review]

**Comment by AOx0** [comment]
Thank you Vito :D

**Pull request merged**
